src/v/resource_mgmt/tests/cpu_profiler_test.cc

```diff
@@ -16,30 +16,125 @@
+        co_await ss::coroutine::maybe_yield();
+    }
+}
+} // namespace
 
+using namespace std::literals;
+
+SEASTAR_THREAD_TEST_CASE(test_cpu_profiler) {
     resources::cpu_profiler cp(
       config::mock_binding(true), config::mock_binding(2ms));
     cp.start().get();
 
     // The profiler service will request samples from seastar every
     // 256ms since the sample rate is 2ms. So we need to be running
     // for at least that long to ensure the service pulls in samples.
-    auto end_time = ss::lowres_clock::now() + 256ms + 10ms;
-    while (ss::lowres_clock::now() < end_time) {
-        // yield to allow timer to trigger and lowres_clock to update
-        ss::thread::maybe_yield();
-    }
+    busy_loop(256ms + 10ms).get();
 
     auto results = cp.shard_results();
     BOOST_TEST(results.samples.size() >= 1);
 }
+
+SEASTAR_TEST_CASE(test_cpu_profiler_enable_override) {
+    // Ensure that overrides to the profiler will enable it and collect samples
+    // for the specified period of time.
+
+    ss::sharded<resources::cpu_profiler> cp;
+    co_await cp.start(config::mock_binding(false), config::mock_binding(2ms));
+    co_await cp.invoke_on_all(&resources::cpu_profiler::start);
+
+    auto wait_ms = 256ms + 10ms;
+    auto [results] = co_await ss::coroutine::all(
+      [&]() {
+          return cp.local().override_and_get_results(wait_ms, std::nullopt);
+      },
+      [&]() { return busy_loop(wait_ms); });
+
+    BOOST_TEST(results[ss::this_shard_id()].samples.size() >= 1);
+
+    // CPU profiler should be disabled so if we wait some time and collect the
+    // samples again there should be no new samples.
+    co_await busy_loop(wait_ms);
+
+    auto new_results = cp.local().shard_results();
+    auto old_results = results[ss::this_shard_id()];
+
+    BOOST_REQUIRE(new_results.samples.size() == old_results.samples.size());
+
+    for (int i = 0; i < old_results.samples.size(); i++) {
+        BOOST_REQUIRE(
+          new_results.samples[i].occurrences
+          == old_results.samples[i].occurrences);
+        BOOST_REQUIRE(
+          new_results.samples[i].user_backtrace
+          == old_results.samples[i].user_backtrace);
+    }
+
+    co_await cp.stop();
+}
+
+SEASTAR_TEST_CASE(test_cpu_profiler_enable_override_nested) {
+    // Ensure that in cases with multiple on-going overrides that the shortest
+    // override doesn't prematurely disable the profiler.
+
+    ss::sharded<resources::cpu_profiler> cp;
+    co_await cp.start(config::mock_binding(false), config::mock_binding(10ms));
+    co_await cp.invoke_on_all(&resources::cpu_profiler::start);
+
+    auto wait_ms = 1s;
+    auto [results_p, results_10p] = co_await ss::coroutine::all(
+      [&]() {
+          return cp.local().override_and_get_results(wait_ms, std::nullopt);
+      },
+
+      [&]() {
+          return cp.local().override_and_get_results(
+            10 * wait_ms, std::nullopt);
+      },
+      [&]() { return busy_loop(10 * wait_ms); });
```

Comment:
I guess by doing the two different profiles in a `when_all` you are not exactly checking that they don't interfere. The shorter one could still make the long one stop?

But I don't think this is a super common scenario so not sure how much time you want to spend on this.

Reply:
I later check that the longer one received more samples than the shorter one. This is hopefully a sufficient check to ensure the short one didn't cause the longer one to stop early.